src/liballoc/arc.rs

```diff
@@ -555,8 +555,9 @@ impl<T: ?Sized> Drop for Arc<T> {
         // more than once (but it is guaranteed to be zeroed after the first if
         // it's run more than once)
         let ptr = *self._ptr;
-        // if ptr.is_null() { return }
-        if ptr as *mut u8 as usize == 0 || ptr as *mut u8 as usize == mem::POST_DROP_USIZE {
+        let thin = ptr as *const ();
+
+        if !thin.is_null() && thin as usize != mem::POST_DROP_USIZE {
```

Comment:
Why did you invert the condition?

It should either be `if thin.is_null() || thin as usizee == mem::POST_DROP_USIZE` or `if !(!thin.is_null() && thin as usize != mem::POST_DROP_USIZE)`. What you wrote is not equivalent to the old code.